{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from collections import Counter\n",
    "from typing import List, Dict\n",
    "import numpy as np\n",
    "import numpy as np\n",
    "from typing import List, Dict\n",
    "from collections import Counter\n",
    "from numpy.linalg import norm\n",
    "\n",
    "START_CODON = \"ATG\"\n",
    "STOP_CODONS = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "FASTA_DIRECTORY = \"viruses/data\"\n",
    "file_prefixes = [\"bacterial\", \"mamalian\"]\n",
    "amino_acids = \"ACDEFGHIKLMNPQRSTVWY\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Virus:\n",
    "    def __init__(self, fileName, name, sequence, orfs):\n",
    "        self.fileName = fileName\n",
    "        self.name = name\n",
    "        self.sequence = sequence \n",
    "        self.orfs = orfs \n",
    "        self.protein_sequences = []  \n",
    "        self.proteinsDicodones = [] \n",
    "        self.codon_frequency = [] \n",
    "        self.dicodon_frequency = []\n",
    "        self.amino_acid_frequencies = {}\n",
    "\n",
    "    def calculate_frequencies(self):\n",
    "        combined_sequence = ''.join(self.protein_sequences)\n",
    "        self.amino_acid_frequencies = self.calculate_amino_acid_frequencies(combined_sequence)\n",
    "        self.dicodon_frequency = self.calculate_dicodon_frequencies(combined_sequence)\n",
    "\n",
    "    @staticmethod\n",
    "    def calculate_amino_acid_frequencies(sequence: str) -> Dict[str, float]:\n",
    "        amino_acid_count = Counter(sequence)\n",
    "        total_amino_acids = sum(amino_acid_count.values())\n",
    "        return {amino_acid: count / total_amino_acids for amino_acid, count in amino_acid_count.items()}\n",
    "\n",
    "    @staticmethod\n",
    "    def calculate_dicodon_frequencies(sequence: str) -> Dict[str, float]:\n",
    "        dicodons = [sequence[i:i + 2] for i in range(len(sequence) - 1)]\n",
    "        dicodon_count = Counter(dicodons)\n",
    "        total_dicodons = sum(dicodon_count.values())\n",
    "        return {dicodon: count / total_dicodons for dicodon, count in dicodon_count.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_sequence(seq):\n",
    "    \n",
    "    if START_CODON not in seq:\n",
    "        return False, \"No start codon found\"\n",
    "    \n",
    "    if seq[-3:] not in STOP_CODONS:\n",
    "        return False, \"No valid stop codon found\"\n",
    "    \n",
    "    if len(seq) % 3 != 0:\n",
    "        return False, \"Sequence length is not divisible by 3\"\n",
    "    \n",
    "    if len(seq) < 100:\n",
    "        return False, \"Sequence length is less than 100 bp\"\n",
    "    \n",
    "    return True, \"Sequence is valid\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_orfs(sequence):\n",
    "    orfs = []\n",
    "    seq_len = len(sequence)\n",
    "    \n",
    "    for strand, nuc_seq in [(+1, sequence), (-1, sequence.reverse_complement())]:\n",
    "        for frame in range(3):\n",
    "            start_pos = None\n",
    "            for i in range(frame, seq_len - 2, 3):\n",
    "                codon = str(nuc_seq[i:i+3])\n",
    "                if codon == START_CODON:\n",
    "                    start_pos = i\n",
    "                elif codon in STOP_CODONS and start_pos is not None:\n",
    "                    stop_pos = i + 3\n",
    "                    if (stop_pos - start_pos) >= 100:\n",
    "                        orfs.append(str(nuc_seq[start_pos:stop_pos]))\n",
    "                    start_pos = None\n",
    "    return orfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_fasta(file_path):\n",
    "    all_orfs = []\n",
    "    for record in SeqIO.parse(file_path, \"fasta\"):\n",
    "        orfs = find_orfs(record.seq)\n",
    "        all_orfs.extend(orfs)\n",
    "    return all_orfs, record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_files(fasta_directory, file_prefixes):\n",
    "    for prefix in file_prefixes:\n",
    "        file_path = f\"{fasta_directory}/{prefix}.fasta\"\n",
    "        orfs, record = process_fasta(file_path)\n",
    "        virus = Virus(fileName=f\"{prefix}.fasta\", name=record.name, sequence=record.seq, orfs=orfs)\n",
    "    return virus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_to_protein(virus):\n",
    "    protein_sequences = []\n",
    "    proteinsDicodones = []\n",
    "\n",
    "    for sequence  in virus.orfs:\n",
    "        protein = Seq(sequence).translate()\n",
    "        protein_sequences.append(str(protein))\n",
    "\n",
    "        dicodons = [str(protein[i:i+2]) for i in range(len(protein)-1)]\n",
    "        proteinsDicodones.append(dicodons[:-1])\n",
    "    return protein_sequences, proteinsDicodones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_frequencies(viruses: List[Virus], frequency_type: str) -> Dict[str, int]:\n",
    "    unique_frequencies = set()\n",
    "    \n",
    "    for virus in viruses:\n",
    "        if frequency_type == 'amino_acid':\n",
    "            unique_frequencies.update(virus.amino_acid_frequencies.keys())\n",
    "        elif frequency_type == 'dicodon':\n",
    "            unique_frequencies.update(virus.dicodon_frequency.keys())\n",
    "        else:\n",
    "            raise ValueError(f\"Unknown frequency type: {frequency_type}\")\n",
    "    \n",
    "    return {item: idx for idx, item in enumerate(sorted(unique_frequencies))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_distance_matrix(viruses: List[Virus], frequency_type: str = 'amino_acid') -> np.ndarray:\n",
    "    num_viruses = len(viruses)\n",
    "    distance_matrix = np.zeros((num_viruses, num_viruses))\n",
    "\n",
    "    unique_frequencies = get_unique_frequencies(viruses, frequency_type)\n",
    "\n",
    "    for i in range(num_viruses):\n",
    "        frequency_vector_i = np.zeros(len(unique_frequencies))\n",
    "        \n",
    "        if frequency_type == 'amino_acid':\n",
    "            for amino_acid, freq in viruses[i].amino_acid_frequencies.items():\n",
    "                frequency_vector_i[unique_frequencies[amino_acid]] = freq\n",
    "        elif frequency_type == 'dicodon':\n",
    "            for dicodon, freq in viruses[i].dicodon_frequency.items():\n",
    "                frequency_vector_i[unique_frequencies[dicodon]] = freq\n",
    "\n",
    "        for j in range(i, num_viruses):\n",
    "            frequency_vector_j = np.zeros(len(unique_frequencies))\n",
    "            if frequency_type == 'amino_acid':\n",
    "                for amino_acid, freq in viruses[j].amino_acid_frequencies.items():\n",
    "                    frequency_vector_j[unique_frequencies[amino_acid]] = freq\n",
    "            elif frequency_type == 'dicodon':\n",
    "                for dicodon, freq in viruses[j].dicodon_frequency.items():\n",
    "                    frequency_vector_j[unique_frequencies[dicodon]] = freq\n",
    "\n",
    "            i_j = frequency_vector_i - frequency_vector_j\n",
    "                        \n",
    "            distance = norm(i_j, ord=2)\n",
    "\n",
    "            distance_matrix[i][j] = distance\n",
    "            distance_matrix[j][i] = distance\n",
    "\n",
    "    return distance_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_phylip_format(viruses: List[Virus], distance_matrix: np.ndarray, file_name: str):\n",
    "    num_viruses = len(viruses)\n",
    "\n",
    "    if distance_matrix.shape[0] != num_viruses or distance_matrix.shape[1] != num_viruses:\n",
    "        raise ValueError(\"Distance matrix dimensions do not match the number of viruses\")\n",
    "\n",
    "    phylip_lines = []\n",
    "    \n",
    "    phylip_lines.append(f\"{num_viruses}\")\n",
    "    \n",
    "    for i in range(num_viruses):\n",
    "        virus_name = viruses[i].name\n",
    "        distances = [f\"{distance_matrix[i][j]:8.4f}\" for j in range(num_viruses)] \n",
    "        \n",
    "        phylip_lines.append(f\"{virus_name:<15}{' '.join(distances)}\")\n",
    "\n",
    "    with open(file_name, 'w') as file:\n",
    "        file.write(\"\\n\".join(phylip_lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Surinkti duomenis ir start end poras\n",
    "viruses = []\n",
    "for i in range(1, 5):\n",
    "    for prefix in file_prefixes:\n",
    "        virus = process_files(FASTA_DIRECTORY, [f\"{prefix}{i}\"])\n",
    "        viruses.append(virus)\n",
    "\n",
    "# Konvertuoti sekas į baltymo kodonus ir dikodonus\n",
    "for virus in viruses:\n",
    "    virus.protein_sequences, virus.proteinsDicodones = translate_to_protein(virus)\n",
    "\n",
    "# Apskaičiuoti baltymo ir dikodonų dažnius\n",
    "for virus in viruses:\n",
    "    virus.calculate_frequencies()\n",
    "\n",
    "# Apskaičiuoti atstumų matricas\n",
    "amino_acid_distance_matrix = calculate_distance_matrix(viruses, frequency_type='amino_acid')\n",
    "dicodon_distance_matrix = calculate_distance_matrix(viruses, frequency_type='dicodon')\n",
    "\n",
    "# Spausdinti matricas, jei reikia\n",
    "if False:\n",
    "    print(\"Amino Acid Distance Matrix:\")\n",
    "    print(amino_acid_distance_matrix)\n",
    "\n",
    "    print(\"Dicodon Distance Matrix:\")\n",
    "    print(dicodon_distance_matrix)\n",
    "\n",
    "# Sukurti ir išsaugoti atstumo matricas phylip formatu\n",
    "file_name = \"codon viruses_distance_matrix.phy\"\n",
    "write_phylip_format(viruses, amino_acid_distance_matrix, file_name)\n",
    "file_name = \"dicodon viruses_distance_matrix.phy\"\n",
    "write_phylip_format(viruses, dicodon_distance_matrix, file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 5 labiausiai varijuojantys kodonai:\n",
      "R: 0.02376735004204152\n",
      "I: 0.020765823658179663\n",
      "K: 0.020472715049806457\n",
      "A: 0.01492226505004192\n",
      "N: 0.013723334338643492\n",
      "\n",
      "Top 5 labiausiai varijuojantys dikodonai:\n",
      "RR: 0.005642487278072178\n",
      "LL: 0.004438320449881403\n",
      "AA: 0.003428929951903611\n",
      "II: 0.0032532247332899104\n",
      "GG: 0.0030099978834778607\n"
     ]
    }
   ],
   "source": [
    "# Labiausiai varijuojančių kodonų ir dikodonų paieška\n",
    "def find_most_variable_codons_dicodons(viruses, frequency_type='codon', top_n=5):\n",
    "   \n",
    "    all_frequencies = []\n",
    "    if frequency_type == 'codon':\n",
    "        codon_list = list(viruses[0].amino_acid_frequencies.keys()) \n",
    "        for virus in viruses:\n",
    "            codon_freqs = [virus.amino_acid_frequencies.get(codon, 0) for codon in codon_list]\n",
    "            all_frequencies.append(codon_freqs)\n",
    "    elif frequency_type == 'dicodon':\n",
    "        codon_list = list(viruses[0].dicodon_frequency.keys())\n",
    "        for virus in viruses:\n",
    "            dicodon_freqs = [virus.dicodon_frequency.get(dicodon, 0) for dicodon in codon_list]\n",
    "            all_frequencies.append(dicodon_freqs)\n",
    "\n",
    "    all_frequencies = np.array(all_frequencies)\n",
    "    std_devs = np.std(all_frequencies, axis=0)\n",
    "    sorted_indices = np.argsort(std_devs)[::-1]\n",
    "    most_variable = [(codon_list[i], std_devs[i]) for i in sorted_indices[:top_n]]\n",
    "\n",
    "    return most_variable\n",
    "\n",
    "top_variable_codons = find_most_variable_codons_dicodons(viruses, frequency_type='codon', top_n=5)\n",
    "top_variable_dicodons = find_most_variable_codons_dicodons(viruses, frequency_type='dicodon', top_n=5)\n",
    "\n",
    "print(\"Top 5 labiausiai varijuojantys kodonai:\")\n",
    "for codon, variance in top_variable_codons:\n",
    "    print(f\"{codon}: {variance}\")\n",
    "\n",
    "print(\"\\nTop 5 labiausiai varijuojantys dikodonai:\")\n",
    "for dicodon, variance in top_variable_dicodons:\n",
    "    print(f\"{dicodon}: {variance}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
